src/serac/physics/tests/serac_solid_functional_boundary.cpp

```diff
@@ -0,0 +1,117 @@
+#include <gtest/gtest.h>
+#include "mfem.hpp"
+
+#include "serac/serac_config.hpp"
+#include "serac/mesh/mesh_utils.hpp"
+#include "serac/physics/state/state_manager.hpp"
+#include "serac/physics/solid_functional.hpp"
+#include "serac/physics/materials/solid_functional_material.hpp"
+#include "serac/physics/materials/parameterized_solid_functional_material.hpp"
+#include "serac/infrastructure/initialize.hpp"
+
+namespace serac {
+
+TEST(SolidFunctional, BoundaryCondition)
+{
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  int serial_refinement   = 0;
+  int parallel_refinement = 0;
+
+  constexpr int dim = 2;
+  constexpr int p   = 1;
+
+  // Create DataStore
+  axom::sidre::DataStore datastore;
+  serac::StateManager::initialize(datastore, "solid_functional_boundary condition");
+
+  std::string filename = SERAC_REPO_DIR "/data/meshes/beam-quad.mesh";
+
+  auto mesh = mesh::refineAndDistribute(buildMeshFromFile(filename), serial_refinement, parallel_refinement);
+  serac::StateManager::setMesh(std::move(mesh));
+
+  // define the solver configurations
+  const IterativeSolverOptions default_linear_options = {.rel_tol     = 1.0e-6,
+                                                         .abs_tol     = 1.0e-10,
+                                                         .print_level = 0,
+                                                         .max_iter    = 500,
+                                                         .lin_solver  = LinearSolver::GMRES,
+                                                         .prec        = HypreBoomerAMGPrec{}};
+
+  const NonlinearSolverOptions default_nonlinear_options = {
+      .rel_tol = 1.0e-4, .abs_tol = 1.0e-8, .max_iter = 10, .print_level = 1};
+
+  const typename solid_util::SolverOptions default_static = {default_linear_options, default_nonlinear_options};
+
+  // Construct a functional-based solid mechanics solver
+  SolidFunctional<p, dim> solid_solver(default_static, GeometricNonlinearities::Off, FinalMeshOption::Reference,
+                                       "solid_functional_boundary");
+
+  solid_util::LinearIsotropicSolid<dim> mat(1.0, 1.0, 1.0);
+  solid_solver.setMaterial(mat);
+
+  // Define the function for the Dirichlet boundary conditions
+  auto zero_bc      = [](const mfem::Vector&, mfem::Vector& bc_vec) -> void { bc_vec = 0.0; };
+  auto displaced_bc = [](const mfem::Vector&, mfem::Vector& bc_vec) -> void {
+    bc_vec[0] = 0.0;
+    bc_vec[1] = 0.1;
+  };
+
+  // Set the initial displacement and boundary condition
+  solid_solver.setDisplacementBCs({1}, zero_bc);
+  solid_solver.setDisplacementBCs({2}, displaced_bc);
+  solid_solver.setDisplacement(zero_bc);
+
+  // Finalize the data structures
+  solid_solver.completeSetup();
+
+  // Perform the quasi-static solve
+  double dt = 0.1;
+
+  solid_solver.advanceTimestep(dt);
+  solid_solver.outputState("paraview_output");
+
+  // Check the final displacement norm
+  EXPECT_NEAR(0.171720048, norm(solid_solver.displacement()), 1.0e-6);
+
```

Comment:
I don't want to be too much of a jerk here, but I said earlier that I will not approve any PRs that introduce tests like this. I understand this is a 1-line-of-code change PR, but we can't keep doing this.

Let's start doing basic patch tests or something-- linear elasticity is fine for testing this feature, so we have analytic solutions to compare the entire displacement solution to.

I just glanced at mfem and didn't see any obvious patch test meshes in their repo (?!) so I guess we have to write our own, but it's pretty easy to do.

Reply:
I totally agree. I tried to be a bit better here by actually checking the exact dof values. I agree we need real patch tests.

Reply:
I created an issue so we can track this and start the conversation outside of a bugfix PR.

#754